File: app/src/main/java/ru/buylist/utils/BindingAdapters.java
package ru.buylist.utils;

import androidx.databinding.BindingAdapter;
import android.view.View;
import android.widget.EditText;

import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.floatingactionbutton.FloatingActionButton;

import java.util.List;

import ru.buylist.buy_list.BuyListAdapter;
import ru.buylist.data.entity.Item;
import ru.buylist.pattern_list.PatternListAdapter;


public class BindingAdapters {

    //  задает видимость слою создания нового листа
    @BindingAdapter("visibleGone")
    public static void showHide(View view, boolean show) {
        view.setVisibility(show ? View.VISIBLE : View.GONE);
    }

    // задает видимость recyclerView
    @BindingAdapter("recyclerVisible")
    public static void recyclerLoad(View view, boolean show) {
        view.setVisibility(show ? View.VISIBLE : View.GONE);
    }

    // устанавливает список товаров
    @BindingAdapter("app:buyItems")
    public static void setItems(RecyclerView recyclerView, List<Item> items) {
        BuyListAdapter adapter = (BuyListAdapter) recyclerView.getAdapter();
        if (adapter != null) {
            adapter.setItems(items);
        }
    }

    // устанавливает список товаров в шаблоне
    @BindingAdapter("app:patternItems")
    public static void setPatternItems(RecyclerView recyclerView, List<Item> items) {
        PatternListAdapter adapter = (PatternListAdapter) recyclerView.getAdapter();
        if (adapter != null) {
            adapter.setItems(items);
        }
    }

    // Скрытие / отображение fab с анимацией
    @BindingAdapter("fabVisibility")
    public static void setFabVisibility(FloatingActionButton button, boolean isShown) {
        if (!isShown) {
            button.hide();
        } else {
            button.show();
        }
    }

    @BindingAdapter("requestFocus")
    public static void setFocus(EditText field, boolean isFocusable) {
        if (isFocusable) {
            field.requestFocus();
        } else {
            field.clearFocus();
        }
    }
}

File: app/src/main/java/ru/buylist/presentation/adapters/BindingAdapters.kt
package ru.buylist.presentation.adapters

import android.annotation.SuppressLint
import android.widget.EditText
import androidx.databinding.BindingAdapter
import androidx.databinding.adapters.TextViewBindingAdapter
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.floatingactionbutton.FloatingActionButton
import ru.buylist.data.entity.wrappers.*
import ru.buylist.utils.hideKeyboard
import ru.buylist.utils.showKeyboard

object BindingAdapters {

    @BindingAdapter("app:items")
    @JvmStatic
    fun setBuyLists(recycler: RecyclerView, items: List<BuyListWrapper>) {
        with(recycler.adapter as BuyListAdapter) {
            list = items
        }
    }

    @BindingAdapter("app:items")
    @JvmStatic
    fun setPatterns(recycler: RecyclerView, patterns: List<PatternWrapper>) {
        with(recycler.adapter as PatternsAdapter) {
            list = patterns
        }
    }

    @BindingAdapter("app:items")
    @JvmStatic
    fun setRecipes(recycler: RecyclerView, recipes: List<RecipeWrapper>?) {
        recipes?.let {
            (recycler.adapter as RecipesAdapter).submitList(recipes)
        }
    }

    @BindingAdapter("app:items")
    @JvmStatic
    fun setItems(recycler: RecyclerView,
                 items: List<ItemWrapper>) {
        with(recycler.adapter as BuyListDetailAdapter) {
            wrappedItems = items
        }
    }

    @BindingAdapter("app:patternItems")
    @JvmStatic
    fun setPatternItems(recycler: RecyclerView,
                        items: List<ItemWrapper>) {
        with(recycler.adapter as PatternDetailAdapter) {
            wrappedItems = items
        }
    }

    @BindingAdapter("app:items")
    @JvmStatic
    fun setCircles(recycler: RecyclerView, circles: List<CircleWrapper>) {
        with(recycler.adapter as CirclesAdapter) {
            list = circles
        }
    }

    @BindingAdapter("app:fabVisibility")
    @JvmStatic
    fun setFabVisibility(fab: FloatingActionButton, isShown: Boolean) {
        if (isShown) {
            fab.show()
        } else {
            fab.hide()
        }
    }

    @SuppressLint("RestrictedApi")
    @BindingAdapter("android:text")
    @JvmStatic
    fun setSelection(field: EditText, text: String?) {
        TextViewBindingAdapter.setText(field, text)

        if (text == null || text.isEmpty()) return
        field.setSelection(text.trim().length)
    }

    @BindingAdapter("app:requestFocus")
    @JvmStatic
    fun requestFocus(field: EditText, requestFocus: Boolean) {
        if (requestFocus) {
            field.showKeyboard()
        } else {
            field.hideKeyboard()
        }
    }
}